Added Audiosource & AudioListener component

AtlantiaKing opened the pull request:
Added IAudioClip (used by user) and FmodAudioClip (used internal) classes
Added AudioSource component with support for 3D & 2D sound, pausing, muting, looping and PlayOnAwake
Added AudioListener component for correct 3D sounds
Added AudioTester component that can call pause, resume, mute, unmute and play on an audiosource using input

AtlantiaKing committed:
Added AudioClip, AudioSource and AudioListener

Added IAudioClip interface with FmodAudioClip as a derived version of the interface
Added clip loading and playing to the IAudioSystem and FmodAudioSystem
Added AudioSource component
Added AudioListener component
Renamed Systems folder to FMOD

AtlantiaKing committed:
Added more setter functions to the AudioSource

Added SetLooping method to the IAudioSystem and FModAudioSystem classes
Added Set3DDistances, Set3DVolume, Set3DData and SetLooping methods to the AudioSource component
Added IsPlaying method to the AudioSource component
Fixed runtime error when closing the engine while a sound is playing

AtlantiaKing committed:
Added support for async load/play

Previously the PlaySound call would fail when playing a async loaded sound because it was "Not Ready" (which means it was  still loading)
Sounds now get queued if they failed to load because of an NotReady error, any other error will still throw a runtime exception.
Load on play AudioSources not also load sounds in async

AtlantiaKing committed:
Added muting & pausing

Added Pause, Resume, Mute and Unmute methods to the AudioSource component
Added Stop call when calling play on an audiosource that is already playing
Added AudioTest component that controls the audiosource with input

LeeVangraefschepe commented on Audio/HelperObjects/ChannelData.h:
```diff
@@ -1,11 +1,19 @@
 {
 	struct ChannelData final
 	{
+		~ChannelData()
```
Unintentional behaviour when a copy is made of the struct.

LeeVangraefschepe commented on LeapEngine/CMakeLists.txt:
```diff
@@ -35,6 +35,6 @@ add_library(LeapEngine
     "Components/RenderComponents/UIComponents/Button.h" 
     "Components/RenderComponents/UIComponents/Button.cpp" 
-    "Components/RenderComponents/UIComponents/ICanvasElement.h")
+    "Components/RenderComponents/UIComponents/ICanvasElement.h" "Components/Audio/AudioSource.h" "Components/Audio/AudioSource.cpp" "Components/Audio/AudioListener.h" "Components/Audio/AudioListener.cpp")
```
Keep the CMake clean

LeeVangraefschepe commented on LeapEngine/Components/Audio/AudioListener.h:
```diff
@@ -0,0 +1,21 @@
+		AudioListener& operator=(AudioListener&& other) = delete;
+
+	protected:
+		virtual void Update() override;
```
Why protected? Use private.

LeeVangraefschepe commented on LeapEngine/Components/Audio/AudioSource.h:
```diff
@@ -0,0 +1,73 @@
+
+		bool IsPlaying() const;
+
+	protected:
```
Set private again

LeeVangraefschepe commented on UnnamedAdventureGame/Components/AudioTester.h:
```diff
@@ -0,0 +1,23 @@
+	{
+	public:
+
+	protected:
```
Private :D

LeeVangraefschepe reviewed (approved):
Check comments for changes. Overall clean header files and good naming.

AtlantiaKing committed:
Audio cleanup & fixed ChannelData copy issue

Fixed ChannelData being copied and stopCallback being called without reason
Added const correctness of functions inside the FmodAudioSystem pImpl
Moved component function overrides to private instead of protected inside the AudioListener, AudioSource and AudioTester components
Cleaned up CMakeLists of the Core Leap Engine
